src/build_loss_objective.jl

```diff
@@ -49,9 +65,18 @@ function build_loss_objective(prob::DiffEqBase.DEProblem,alg,loss,regularization
   if mpg_autodiff
     gcfg = ForwardDiff.GradientConfig(cost_function, autodiff_prototype, autodiff_chunk)
     g! = (x, out) -> ForwardDiff.gradient!(out, cost_function, x, gcfg)
+  elseif lsa_gradient 
+    function g!(x,out)
+      sol_,sens = diffeq_sen_full(prob.f,prob.u0,prob.tspan,x,loss.t;alg=alg)
+      l2lossgradient!(out,sol_,loss.data,sens,length(prob.p))
+    end
+  elseif adjsa_gradient
+    g! = (x,out) -> diffeq_sen_l2!(out,prob.f,prob.u0,prob.tspan,x,loss.t,loss.data,alg)
   else
     g! = (x, out) -> Calculus.finite_difference!(cost_function,x,out,:central)
   end
+
+
   cost_function2 = function (p,grad)
```

Comment:
you need to change this for the adjoint to reduce a solve.

Reply:
Can you explain what you mean?

Reply:
We could finish this up once you can respond here?

Reply:
https://github.com/JuliaOpt/NLopt.jl#tutorial notice that the gradient and the value computation are mixed together here. In the adjoint calculation, the forward solution is required, so that can be used instead of re-solving.

Reply:
That looks NLopt specific though? https://github.com/JuliaDiffEq/DiffEqParamEstim.jl/pull/98/files/51946011756285f36f80c172712a5a0ae6c86460#diff-2fed90081d9e12363b34887426589f36R43 optimisers require a separate function call for the the loss and for gradient evaluation generally?

Reply:
Look at Optim's `fg!`

Reply:
will be addressed in a different PR.